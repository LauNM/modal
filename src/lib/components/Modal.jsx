import React, { useRef, useEffect } from 'react';
import PropTypes from "prop-types";
import "./style.scss";

const CloseIconDefault = () => {
  return (<svg stroke="currentColor" fill="currentColor" strokeWidth="0" viewBox="0 0 24 24" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg">
    <g>
      <path fill="none" d="M0 0h24v24H0z"></path>
      <path d="M12 10.586l4.95-4.95 1.414 1.414-4.95 4.95 4.95 4.95-1.414 1.414-4.95-4.95-4.95 4.95-1.414-1.414 4.95-4.95-4.95-4.95L7.05 5.636z"></path>
    </g>
  </svg>);
}
function Modal (
  {
    isOpen,
    isFullPage= false,
    hasCloseIcon= true,
    closeIcon,
    closeModal,
    divider = false,
    modalBackground,
    header,
    headerBackgroundColor,
    footer,
    footerBackgroundColor,
    ...props
  }
)  {

  const displayModal = isOpen ? "" : "is-closed";
  const modalSize = isFullPage ? "is-full-page" : "";
  const borderDivider = divider ? "solid 1px black" : "";
  const headerStyle = header ? { backgroundColor: headerBackgroundColor, borderBottom: borderDivider } : {};
  const footerStyle = footer ? { backgroundColor: footerBackgroundColor, borderTop: borderDivider } : {};
  const modalContainerRef = useRef();

  useEffect(() => {
    modalContainerRef.current?.focus()
  }, [isOpen])
  return (<div role="dialog" aria-modal="true" aria-labelledby="dialogHeader" aria-describedby="dialogMain"
               data-testid="modal-container" className={ `modal-wrapper ${ displayModal } ${ modalSize }` }>
      <div className="modal" style={ { backgroundColor: modalBackground } }>
        { hasCloseIcon &&
          <button ref={ modalContainerRef } data-testid="close-button" onClick={ closeModal } className="close-button">
            { closeIcon ?? <CloseIconDefault/> }
          </button> }
        <header id="dialogHeader" style={ headerStyle }>{ header }</header>
        { props.children && <div id="dialogMain" data-testid="modal-main" className="modal-main">
          { props.children }
        </div> }
        <footer style={ footerStyle }>{ footer }</footer>
      </div>
    </div>)
}

Modal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  isFullPage: PropTypes.bool,
  hasCloseIcon: PropTypes.bool,
  closeIcon: PropTypes.element,
  closeModal: PropTypes.func,
  divider: PropTypes.bool,
  modalBackground: PropTypes.string,
  header: PropTypes.element,
  headerBackgroundColor: PropTypes.string,
  footer: PropTypes.element,
  footerBackgroundColor: PropTypes.string,
}

export default Modal;