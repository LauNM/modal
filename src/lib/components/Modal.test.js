import {render, screen} from '@testing-library/react'
import '@testing-library/jest-dom'
import Modal from "./Modal";
import renderer from 'react-test-renderer';

test("Modal component should be mounted", async () => {
    render(<Modal isOpen/>)
    expect(screen.getByTestId('modal-container')).toBeTruthy();
} );

test("Modal component is closed", async () => {
  const { container } = render(<Modal isOpen={false}/>)
  expect(container.firstChild).toHaveClass('is-closed')
} );

test("Modal component is opened", async () => {
  const { container } = render(<Modal isOpen/>)
  expect(container.firstChild).not.toHaveClass('is-closed')
} );

test("Modal component is not full page", async () => {
  const { container } = render(<Modal isOpen/>)
  expect(container.firstChild).not.toHaveClass('is-full-page')
} );

test("Modal component is full page", async () => {
  const { container } = render(<Modal isOpen isFullPage/>)
  expect(container.firstChild).toHaveClass('is-full-page')
} );

test("Modal component has close icon", async () => {
  render(<Modal isOpen />)
  expect(screen.getByTestId('close-button')).toBeTruthy();
} );

test("Modal component has not close icon", async () => {
  render(<Modal isOpen hasCloseIcon={false}/>)
  expect(screen.queryByTestId('close-button')).toBeNull();
} );

test("Modal component has header", async () => {
  render(<Modal isOpen header={<h2>Header title</h2>}/>)
  expect(screen.getByText('Header title')).toBeTruthy();;
} );

test("Modal component has footer", async () => {
  render(<Modal isOpen footer={<p>Footer text</p>}/>)
  expect(screen.getByText('Footer text')).toBeTruthy();;
} );

test("Modal component has content container", async () => {
  render(<Modal isOpen><div><h3>Modal content</h3><p>Modal subtitle</p></div></Modal>)
  expect(screen.getByTestId('modal-main')).toBeTruthy();;
} );

test("Modal component has content", async () => {
  render(<Modal isOpen><div><h3>Modal content</h3><p>Modal subtitle</p></div></Modal>)
  expect(screen.getByText('Modal content')).toBeTruthy();;
  expect(screen.getByText('Modal subtitle')).toBeTruthy();;
} );

it('simple modal renders correctly', () => {
    const tree = renderer
      .create(<Modal isOpen/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

it('modal with content renders correctly', () => {
  const tree = renderer
    .create(<Modal isOpen><div><h3>Modal content</h3><p>Modal subtitle</p></div></Modal>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
