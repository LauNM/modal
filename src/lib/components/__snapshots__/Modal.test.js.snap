// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal with content renders correctly 1`] = `
<div
  className="modal-wrapper  "
  data-testid="modal-container"
>
  <div
    className="modal"
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
  >
    <span
      className="close-button"
      data-testid="close-button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 10.586l4.95-4.95 1.414 1.414-4.95 4.95 4.95 4.95-1.414 1.414-4.95-4.95-4.95 4.95-1.414-1.414 4.95-4.95-4.95-4.95L7.05 5.636z"
          />
        </g>
      </svg>
    </span>
    <header
      style={Object {}}
    />
    <div
      className="modal-main"
      data-testid="modal-main"
    >
      <div>
        <h3>
          Modal content
        </h3>
        <p>
          Modal subtitle
        </p>
      </div>
    </div>
    <footer
      style={Object {}}
    />
  </div>
</div>
`;

exports[`simple modal renders correctly 1`] = `
<div
  className="modal-wrapper  "
  data-testid="modal-container"
>
  <div
    className="modal"
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
  >
    <span
      className="close-button"
      data-testid="close-button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 10.586l4.95-4.95 1.414 1.414-4.95 4.95 4.95 4.95-1.414 1.414-4.95-4.95-4.95 4.95-1.414-1.414 4.95-4.95-4.95-4.95L7.05 5.636z"
          />
        </g>
      </svg>
    </span>
    <header
      style={Object {}}
    />
    <footer
      style={Object {}}
    />
  </div>
</div>
`;
